clouddrift/wavelet.py

```diff
@@ -0,0 +1,534 @@
+    wavefft_ = np.fft.fft(wave_)
+    om = 2 * np.pi * np.linspace(0, 1 - 1 / time_length_, time_length_)
+    if time_length_ % 2 == 0:
+        wavefft_ = (
+            wavefft_ * np.exp(1j * -om * (time_length_ + 1) / 2) * np.sign(np.pi - om)
+        )
+    else:
+        wavefft_ = wavefft_ * np.exp(1j * -om * (time_length_ + 1) / 2)
+
+    # here I should be able to automate the tiling without assuming extra dimensions of wave
+    X_ = np.tile(
+        np.expand_dims(np.fft.fft(x_), (-3, -2)),
+        (1, order_length, freq_length, 1),
+    )
+    # finally the transform
+    wtx = np.fft.ifft(X_ * np.conj(wavefft_))
+    # return central part
+    wtx = wtx[..., index]
+
+    # if I comment this the complex test passes but not the boundary test
+    # and vice versa
+    # I prefer the complex test to pass for now
+    # not sure I understand if this is needed
+    # if np.all(np.isreal(x_)):
+    #    wtx = np.real(wtx)
+    # should we squeeze wtx? probably
+    wtx = np.squeeze(wtx)
+
+    return wtx
+
+
+def morsewave(
+    n: int,
+    ga: float,
+    be: float,
+    rad_freq: np.ndarray,
+    order: Optional[int] = 1,
+    norm: Optional[str] = "bandpass",
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Generalized Morse wavelets of Olhede and Walden (2002).
+
+    Parameters
+    ----------
+    n: int
+       Length of the wavelet.
+    ga: float
+       Gamma parameter of the wavelet.
+    be: float
+       Beta parameter of the wavelet.
+    rad_freq: np.ndarray
+       The radian frequencies at which the Fourier transform of the wavelets
+       reach their maximum amplitudes. rad_freq is between 0 and 2 * np.pi * 0.5,
+       the normalized Nyquist radian frequency.
+    order: int
+        Wavelet order, default is 1.
+    norm:  str, optional
+       Normalization for the wavelets. Default is "bandpass".
+       "bandpass" uses "bandpass normalization", meaning that the FFT of the wavelet
+       has a peak value of 2 for all frequencies rad_freq. "energy" uses the unit
+       energy normalization. The time-domain wavelet energy np.sum(np.abs(wave)**2)
+       is then always unity.
+
+    Returns
+    -------
+    wave : np.ndarray
+        Time-domain wavelets. wave will be of shape (n,np.size(rad_freq),k).
+    wavefft: np.ndarray
+        Frequency-domain wavelets. wavefft will be of shape (n,np.size(rad_freq),k).
+
+    Examples
+    --------
+    To write.
+
+    See Also
+    --------
+    :func:`wavetrans`, `morsefreq`
+    """
+    # add a test for rad_freq being a numpy array
+    # initialization
+    wave = np.zeros((n, order, len(rad_freq)), dtype="cdouble")
+    wavefft = np.zeros((n, order, len(rad_freq)), dtype="cdouble")
+
+    # call to morsewave take only ga and be as float, no array
+    fo, _, _ = morsefreq(ga, be)
+    for i in range(0, len(rad_freq)):
+        wave_tmp = np.zeros((n, order), dtype="cdouble")
+        wavefft_tmp = np.zeros((n, order), dtype="cdouble")
```

Comment:
```suggestion
    wave = np.zeros((n, order, len(rad_freq)), dtype=np.complex128)
    wavefft = np.zeros((n, order, len(rad_freq)), dtype=np.complex128)

    # call to morsewave take only ga and be as float, no array
    fo, _, _ = morsefreq(ga, be)
    for i in range(0, len(rad_freq)):
        wave_tmp = np.zeros((n, order), dtype=np.complex128)
        wavefft_tmp = np.zeros((n, order), dtype=np.complex128)
```

Reply:
I think that complex128 is an alias which is platform-dependent? cdouble seems to be norm? see https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.double